File: app/agent_contract.py
"""Пакет инструкций агентов (app/prompts/agent-contract).

Единый источник текстов ролей и правил для подписочных CLI (Claude Code,
Codex) и Python-пути. Тот же пакет читает desktop/services/agent-contract.mjs;
адаптеры собирают «роль + правило инструментов + правило вывода» одинаково,
различается только транспорт. Версия пакета пишется в generationLog и в
transport-метаданные ответов.
"""
from __future__ import annotations

import json
import os
import re
import threading
from dataclasses import dataclass
from pathlib import Path

APP_DIR = Path(os.environ.get("DESIGNDNA_APP_DIR") or Path(__file__).resolve().parent)
DEFAULT_DIR = APP_DIR / "prompts" / "agent-contract"
_VERSION_RE = re.compile(r"^agent-contract/\d+\.\d+$")
_ROLE_RE = re.compile(r"^[a-z_]{1,32}$")
_LOCK = threading.Lock()
_CACHE: dict[str, "AgentContract"] = {}


@dataclass(frozen=True)
class Role:
    name: str
    output: str
    instructions: str


@dataclass(frozen=True)
class AgentContract:
    version: str
    directory: Path
    roles: dict[str, Role]
    tool_rules: dict[str, str]
    output_rules: dict[str, str]
    reminders: dict[str, str]
    effort_levels: tuple[str, ...]
    effort_default: str

    def has_role(self, name: str) -> bool:
        return name in self.roles

    def role(self, name: str) -> Role:
        try:
            return self.roles[name]
        except KeyError:
            raise ValueError(f"Unsupported agent role: {name}") from None

    def tool_rule(self, kind: str) -> str:
        try:
            return self.tool_rules[kind]
        except KeyError:
            raise ValueError(f"Unsupported tool rule: {kind}") from None

    def compose(self, name: str, tool_rule: str) -> str:
        """«Роль + правило инструментов + правило вывода» — как в JS-адаптерах."""
        role = self.role(name)
        parts = [role.instructions, self.tool_rule(tool_rule), self.output_rules.get(role.output, "")]
        return " ".join(part for part in parts if part)

    def reminder(self, name: str) -> str:
        return self.reminders.get(self.role(name).output, "")

    def summary(self) -> dict:
        """Публичное описание для /api/config и UI: без текстов ролей."""
        return {
            "version": self.version,
            "roles": {name: {"output": role.output} for name, role in self.roles.items()},
            "effort": {"levels": list(self.effort_levels), "default": self.effort_default},
        }


def _read(path: Path) -> str:
    return path.read_text(encoding="utf-8").replace("\r\n", "\n")


def load(directory: Path | str | None = None, *, reload: bool = False) -> AgentContract:
    root = Path(directory or DEFAULT_DIR).resolve()
    key = str(root)
    with _LOCK:
        if not reload and key in _CACHE:
            return _CACHE[key]
    raw = json.loads(_read(root / "contract.json"))
    version = str(raw.get("version") or "").strip()
    if not _VERSION_RE.match(version):
        raise ValueError(f"agent-contract: unsupported version {version!r}")
    tool_rules = {str(k): str(v) for k, v in (raw.get("toolRules") or {}).items()}
    output_rules = {str(k): str(v) for k, v in (raw.get("outputRules") or {}).items()}
    reminders = {str(k): str(v) for k, v in (raw.get("reminders") or {}).items()}
    roles: dict[str, Role] = {}
    for name, spec in (raw.get("roles") or {}).items():
        if not _ROLE_RE.match(str(name)):
            raise ValueError(f"agent-contract: invalid role name {name!r}")
        output = str((spec or {}).get("output") or "")
        if output not in output_rules:
            raise ValueError(f"agent-contract: role {name} has unknown output {output!r}")
        file = str((spec or {}).get("file") or "")
        if not file or ".." in file or Path(file).is_absolute():
            raise ValueError(f"agent-contract: role {name} has an invalid file path")
        instructions = _read(root / file).strip()
        if not instructions:
            raise ValueError(f"agent-contract: role {name} has empty instructions")
        roles[str(name)] = Role(str(name), output, instructions)
    effort = raw.get("effort") or {}
    contract = AgentContract(
        version=version,
        directory=root,
        roles=roles,
        tool_rules=tool_rules,
        output_rules=output_rules,
        reminders=reminders,
        effort_levels=tuple(str(level) for level in (effort.get("levels") or ("medium", "high", "max"))),
        effort_default=str(effort.get("default") or "medium"),
    )
    with _LOCK:
        _CACHE[key] = contract
    return contract


def version() -> str:
    return load().version

File: desktop/services/agent-contract.mjs
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

/* Пакет инструкций агентов (app/prompts/agent-contract): единый источник
 * текстов ролей для Claude Code, Codex и Python-пути. Адаптеры не хранят
 * собственных копий профилей — они собирают «роль + правило инструментов +
 * правило вывода» из пакета одинаково, а различается только транспорт.
 * Версия пакета уходит в transport-метаданные ответа, чтобы результат ноды
 * можно было сопоставить с версией инструкций. */

const moduleDirectory = path.dirname(fileURLToPath(import.meta.url));
/* В dev — <repo>/app/prompts/agent-contract; в упакованном приложении app/
 * лежит рядом с desktop-пакетом в resources (см. forge extraResource), и
 * main.mjs передаёт каталог явно через runtimeRoot. */
export const DEFAULT_AGENT_CONTRACT_DIR = path.resolve(moduleDirectory, "..", "..", "app", "prompts", "agent-contract");

const cache = new Map();

function requireObject(value, field) {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`agent-contract: ${field} must be an object`);
  return value;
}

function buildContract(dir, raw, readFile) {
  const version = String(raw.version || "").trim();
  if (!/^agent-contract\/\d+\.\d+$/.test(version)) throw new Error(`agent-contract: unsupported version "${version}"`);
  const toolRules = Object.freeze({ ...requireObject(raw.toolRules, "toolRules") });
  const outputRules = Object.freeze({ ...requireObject(raw.outputRules, "outputRules") });
  const reminders = Object.freeze({ ...requireObject(raw.reminders, "reminders") });
  const roles = {};
  for (const [name, spec] of Object.entries(requireObject(raw.roles, "roles"))) {
    if (!/^[a-z_]{1,32}$/.test(name)) throw new Error(`agent-contract: invalid role name "${name}"`);
    const output = String(spec?.output || "");
    if (!Object.hasOwn(outputRules, output)) throw new Error(`agent-contract: role ${name} has unknown output "${output}"`);
    const file = String(spec?.file || "");
    if (!file || file.includes("..") || path.isAbsolute(file)) throw new Error(`agent-contract: role ${name} has an invalid file path`);
    const instructions = String(readFile(path.join(dir, file), "utf8")).replace(/\r\n/g, "\n").trim();
    if (!instructions) throw new Error(`agent-contract: role ${name} has empty instructions`);
    roles[name] = Object.freeze({ name, output, instructions });
  }
  const effort = Object.freeze({
    levels: Object.freeze([...(raw.effort?.levels || ["medium", "high", "max"])]),
    default: String(raw.effort?.default || "medium"),
  });
  return Object.freeze({
    version,
    dir,
    roles: Object.freeze(roles),
    toolRules,
    outputRules,
    reminders,
    effort,
    hasRole(name) { return Object.hasOwn(roles, String(name)); },
    role(name) {
      const role = roles[String(name)];
      if (!role) throw new Error(`Unsupported agent role: ${name}`);
      return role;
    },
    /** «Роль + правило инструментов + правило вывода» — одинаково для всех провайдеров. */
    composeInstructions(name, toolRule) {
      const role = this.role(name);
      if (!Object.hasOwn(toolRules, String(toolRule))) throw new Error(`Unsupported tool rule: ${toolRule}`);
      return [role.instructions, toolRules[toolRule], outputRules[role.output]].filter(Boolean).join(" ");
    },
    /** Напоминание в конце пользовательского ввода (пусто для текстовых ролей). */
    reminder(name) {
      return reminders[this.role(name).output] || "";
    },
  });
}

export function loadAgentContract(dir = DEFAULT_AGENT_CONTRACT_DIR, { readFile = readFileSync, reload = false } = {}) {
  const key = path.resolve(dir);
  if (!reload && cache.has(key)) return cache.get(key);
  const raw = JSON.parse(String(readFile(path.join(key, "contract.json"), "utf8")));
  const contract = buildContract(key, raw, readFile);
  cache.set(key, contract);
  return contract;
}
